src/crimson/os/seastore/segment_manager/block.cc

```diff
@@ -61,23 +62,49 @@ static write_ertr::future<> do_writev(
   // size, we need to rebuild here
   bl.rebuild_aligned(block_size);
 
-  std::vector<iovec> iov;
-  bl.prepare_iov(&iov);
-  return device.dma_write(
-    offset,
-    std::move(iov)
-  ).handle_exception([](auto e) -> write_ertr::future<size_t> {
-      logger().error(
-	"do_writev: dma_write got error {}",
-	e);
-      return crimson::ct_error::input_output_error::make();
-  }).then(
-    [bl=std::move(bl)/* hold the buf until the end of io */](size_t written)
-	       -> write_ertr::future<> {
-    if (written != bl.length()) {
-      return crimson::ct_error::input_output_error::make();
+  return seastar::do_with(
+    std::vector<std::pair<uint64_t, std::vector<iovec>>>(),
+    std::move(bl),
+    [&device, offset] (auto& iovs, auto& bl) {
+    size_t iovs_i = 0;
+    size_t index = 0;
+    uint64_t off = offset;
+    iovs.resize(bl.get_num_buffers() / IOV_MAX + 1);
+    iovs[iovs_i].second.resize(
+      std::min(bl.get_num_buffers() - IOV_MAX * iovs_i, (size_t)IOV_MAX));
+    iovs[iovs_i].first = off;
+    for (auto& bp : bl.buffers()) {
+      iovs[iovs_i].second[index].iov_base = (void*)bp.c_str();
+      iovs[iovs_i].second[index++].iov_len = bp.length();
+      off += bp.length();
+      if (index > IOV_MAX) {
+	iovs_i++;
+	index = 0;
+	iovs[iovs_i].first = off;
+	iovs[iovs_i].second.resize(
+	  std::min(bl.get_num_buffers() - IOV_MAX * iovs_i, (size_t)IOV_MAX));
+      }
     }
-    return write_ertr::now();
+
+    return write_ertr::parallel_for_each(
+      iovs,
+      [&device, &bl](auto& p) mutable {
+      auto& off = p.first;
+      auto& iov = p.second;
+      return device.dma_write(off, std::move(iov))
+      .handle_exception([](auto e) -> write_ertr::future<size_t> {
+	logger().error(
+	  "do_writev: dma_write got error {}",
+	  e);
+	return crimson::ct_error::input_output_error::make();
+      }).then(
```

Comment:
Why it isn't a `safe_then()`? Looks `handle_exception` returns an errorated future internally.

Reply:
`seastar::future::handle_exception()` only returns `seastar::future<>`

Reply:
Looks strange to me. Seems that will swallow the input_output_error that need to be populated.

Reply:
Should we at least add `// FIXME` here if it is an issue?

Reply:
> To put it another way, in case of dma write exception, the `handle_exception` here will return a `seastar::future<size_t>` holding a `crimson::ct_error::input_output_error` which will make seastar reactor skip the next `seastar::future<size_t>::then()` invocation and be handled by later `crimson::errorator<AllowedErrors...>::handle_error` or `crimson::errorator<AllowedErrors...>::safe_then`.

The reason that `seastar::future<size_t>` can hold a `crimson::ct_error::input_output_error` is that: first, we specialized `seastar::futurize<>` for errorated future in `errorator.h`; second, we made `seastar::future<T>` a friend to `crimson::errorator<AllowedErrors...>::future<T>`

Reply:
IIUC errorator is designed to explicitly pass potential errors through the `safe_then()` chain, so that we will not forget (or even forced by the compiler) to handle errors. It should be a loophole if we can hide an errorated future in the `then()` chain.

If we cannot prevent this usage with `handle_exception()` right now, I think FIXME comment is needed.

cc @rzarzynski

Reply:
> Yeah, I'm also worried that this isn't actually propagating the error. @xxhdx1985126 Can you explain how this works?

I think it is propagating the error.